doc/mgmt/Sonic_mirror.md

```diff
@@ -0,0 +1,356 @@
+# Feature Name
+Layer 3 Mirror Session
+# High Level Design Document
+#### Rev 0.1
+
+# Table of Contents
+  * [List of Tables](#list-of-tables)
+  * [Revision](#revision)
+  * [About This Manual](#about-this-manual)
+  * [Scope](#scope)
+  * [Definition/Abbreviation](#definitionabbreviation)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+# Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 10/28/2019  |   Sucheta Mahara      | Initial version                   |
+
+# About this Manual
+This document provides information about configuring Mirror Sessions and ACL rules for mirrorring on an interface  using the management framework based on current capabilities of SONiC.
+
+# Scope
+
+This document cover mirror session configuration and show commands. It also covers ip mirror Access list configuration and show commands
+.
+
+# Definition/Abbreviation
+
+### Table 1: Abbreviations
+
+
+# 1 Feature Overview
+
+This feature will allow the user to configure mirror session and ip mirror access list on an interface using SONiC Management Framework via CLI, REST or  gNMI. The underlying mirror session DB schema is already provided by mirror session support from SONiC.
+The mirror sessions created can then be associated to ip mirror access list and ACL rules can be configured to mirror selected packet only.
+The mirror access list can then be associated to interfaces for mirroring packets on that interface.
+
+Note: Currently the backend doesn't support changing mirror session parameters once created.
+Also certain parameters are required to create a mirror session as defined in the sonic yang for mirror session.
+
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+Provide management framework support to existing SONiC capabilities with respect to mirror session and ip mirror access list.
+
+### 1.1.2 Configuration and Management Requirements
+
+1. CLI configuration/show support
+2. REST API support
+3. gNMI support
+
+### 1.1.3 Scalability Requirements
+N/A
+
+### 1.1.4 Warm Boot Requirements
+N/A
+
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+Implement mirror session, ip mirror access list, ip mirror access group support using transformer in sonic-mgmt-framework.
```

Comment:
Adding a new ACL type is not extensible. We have to do the same for all other ACL types also like MAC, IPv6, UDF etc. Even when IPv4 ACL itself is enhanced with new Keys, The CLIs have to changed for regular ACLs and mirror ACLs etc. A better approach would be to reuse the existing ACLs configured by the user and reference them by name etc. This way any new ACL types or ACL keys that are added will be automatically available for mirroring also.

Reply:
When creating an ACL table its type needs to be known at start . i.e in sonic ACL_TABLE is created based on the type like ipv4, ipv6 or mirror ( tomorrow may be mac/l2).

There are always different  CLIs  for  different types of access list. Here is an example from cisco and Dell OS10  :--

Router(config)# mac access-list ext macext2 ---------------for MAC
Device(config)# ipv6 access-list inbound--------------------For IPv6
Device(config)# ip access-list extended ---------------------for IPV4

So we have to keep it separate. They can't be differentiated based on names. Moreover they may have different configurable parameters under different access lists.

Reply:
Yes each ACL has a type but I am suggesting is not to create ip-mirror ACL as described in this document (in future ipv6-mirror and mac-mirror) ACL Types. It will be difficult to duplicate ACL Types etc as lets say we want to add support for fragmentation flags for IP ACLs, Manageability code for ip access-group, ip-mirror access-group needs to changed. This kind of duplication of effort should be avoided. There should be only well known ACL types ie MAC, IP and IPv6 visible in CLIs. A new UI construct will use these well known ACL types and reference / attach this ACL to a mirror session. This decouples the ACL with mirror so that any enhancement done for ACL will be automatically extended for mirroring also.